File: chomp.py
# Copyright 2024 (AGPL-3.0-or-later), Miles K. Bertrand et al.

# Why is this module called chomp? Idk, it's just a random word that's vaguely related to processing things (in the case of chomp, it's food but that's beside the point)

import re

def chomp(gabc: str, tags) -> str:

    gabc = gabc.replace(r'<v>\greheightstar</v>','*')
    mode: str | None = None
    if 'mode:' in gabc:
        mode = gabc[gabc.index('mode:') + 5:gabc.index('\n', gabc.index('mode:'))].strip()
        if mode.endswith(';'):
            mode = mode[:-1]

    # Remove all commented text before the beginning of the content
    gabc = '%%\n' + gabc[re.search(r'\([cf]\d\)', gabc).span()[0]:]

    # Replace gregobase's preferred versicle and response with libu-friendly versions
    gabc = gabc.replace('<sp>V/</sp>.', '<v>\\Vbar</v>').replace('<sp>R/</sp>.', '<v>\\Rbar</v>')

    gabc = re.sub('<.?sc>', '', gabc)
    # Remove in-text comments
    gabc = re.sub(r'\[.*?\]', '', gabc)

    if mode:
        gabc = f'mode:{mode};\n{gabc}'
    if 'deus-in-adjutorium' in tags:
        return gabc[0:re.search(r'\(Z\-?\)', gabc).start()]

    elif 'antiphona' in tags:
        euouae = ''
        if '<eu>' in gabc:
            # Notes which define the termination
            euouae = re.sub(r'<.?eu>', '', re.search(r' <eu>.+', gabc).group()).strip()
            # Gabc without the euouae
            gabc = gabc[:gabc.index('<eu>')]

        if not 'paschalis' in tags and ' <i>T. P.</i>' in gabc:
            gabc = gabc[:gabc.index(' <i>T. P.</i>')]

        if not 'septuagesima' in tags and '<i>Post Septuag.</i>' in gabc:
            gabc = gabc[:gabc.index('<i>Post Septuag.</i>')]

        if 'intonata' in tags:
            gabc = gabc[:gabc.index('*')] + '(::)' + euouae
        elif {'commemoratio', 'repetita', 'suffragium'}.isdisjoint(tags):
            gabc = gabc + euouae
        else:
            gabc = gabc.replace('*','')[gabc.index('\n') + 1:]
            firstsyllable = re.search(r'\w+\(', gabc).group()
            gabc = gabc[:gabc.index('(')].capitalize() + gabc[gabc.index('('):].replace(firstsyllable, firstsyllable.capitalize())

        if 'repetita' in tags:
            gabc = 'initial-style:0;\n' + gabc
        else:
            gabc = 'initial-style:1;\n' + gabc
        return gabc

    elif 'responsorium-breve' in tags:
        clef = gabc[:gabc.index(')') + 1]
        incipit = re.search(r'\(.\d\)\s(.+?)\s\*', gabc).group(1)
        response = re.search(r'\*\(\W\)\s(.+?)\s\(::\)', gabc).group(1)
        verses = [''.join(i) if type(i) is tuple else i for i in re.findall(r'<v>\\Vbar<\/v>\.?(\(::\))?\s(.+?)\s\*?\(::\)', gabc)]
        verse = verses[0]
        gloria = verses[1]

        return f'{clef} {incipit} *(;) {response} (::) <v>\\Rbar</v> {incipit} (;) {response} (::) <v>\\Vbar</v> {verse} *(;) {response} (::) <v>\\Vbar</v> {gloria} (::) <v>\\Rbar</v> {incipit} (;) {response} (::)'
    else:
        return gabc

File: frontend/rite-renderer/chomp.js
// Copyright 2026 (AGPL-3.0-or-later), Miles K. Bertrand et al.

import {getPsalmTone} from './psalmify.js';

export function chomp(gabc, tags, resources) {
	gabc = gabc.replace('<v>\\greheightstar</v>', '*');

	let clef;
	let mode = gabc.match(/mode:(.+?)(?:;|\n)/);
	if (mode) {
		mode = mode[1];
	}

	// Remove commented text falling before content
	gabc = gabc.substring(gabc.search(/\n\((.+?)\)/));
	gabc = gabc.replaceAll('<sp>V/</sp>.', '<v>\\Vbar</v>')
		.replaceAll(/<sp>R\/<\/sp>.?/g, '<v>\\Rbar</v>')
		.replaceAll(/<.?sc>/g, '')
		.replaceAll(/\[.*?\]/g, '')
		.replaceAll(/(\(.+?)(\|.+?)(\))/g, '$1$3')
		.replaceAll('<sp>*</sp>', '*')
		.replace(/<c>.+?<\/c>/, '')
		.replaceAll(/<e>(.+?)<\/e>/g, '<i>$1</i>');

	let gabcdata = '';
	if (mode) {
		gabcdata = 'mode:' + mode + ';\n%%\n';
	} else {
		gabcdata = '%%\n';
	}

	// Make sure asterisks are formatted right
	gabc = gabc.replace(/(\([,:;]+?\))\s*?\*\s/, '*$1 ');


  if (tags.has('hymnus') && !tags.has('te-deum')) {
    gabc = gabc.match(/([\s\S]+?)(?:\d\.)?\(::\)/)[1] + '(::)';
  }
	if (tags.has('deus-in-adjutorium')) {
		gabc = gabc.substring(0, gabc.search(/\(Z\-?\)/));

	} else if (tags.has('antiphona')) {
    clef = gabc.match(/(?:^|%)\((.+?)\)/m)
    let euouae = '';
    if (clef) {
      // If clef has a middle letter (very very rare) remove it.
      clef = '' + clef[1][0] + clef[1].at(-1);
      let tone = getPsalmTone(mode, clef, resources);
      if (tone) {
        let ending = tone.terminatio.filter(syl => !syl.includes('r'));
        while (ending.length != 6) {
          ending.unshift(tone.tenor.at(-1));
        }
        euouae = ['E', 'u', 'o', 'u', 'a', 'e.'].map((c, i) => `${c}(${ending[i]})`).join(' ') + ' (::)';
      }
    }

		if (gabc.includes('<i>T. P.</i>')) {
			if (tags.has('in-tempore-paschali')) {
				gabc = gabc.replace('<i>T. P.</i>', '');
			} else {
				gabc = gabc.substring(0, gabc.indexOf('<i>T. P.</i>')).trim();
			}
		}
		if (!tags.has('in-tempore-septuagesimae') && gabc.includes('<i>Post Septuag.</i>')) {
			gabc = gabc.substring(0, gabc.indexOf('<i>Post Septuag.</i>')).trim();
		}
		if (tags.has('intonata')) {
			gabc = gabc.substring(0, gabc.indexOf('*')) + '(::) ' + euouae;
		} else if (tags.has('pars')) {
      gabc = gabc.trim().replace(/^(?:\*\s)?\(.?\)/, '');
      if (clef) {
        gabc = `(${clef}) ` + gabc;
      }
			gabcdata = '%%\n';
		} else if (tags.has('repetita')) {
			gabc = gabc.replace('*', '');
			gabcdata = '%%\n';
		} else if (!(tags.has('commemoratio') || tags.has('suffragium'))) {
			gabc = gabc + euouae;
		}
		
		gabcdata = (tags.has('repetita') || tags.has('pars') ? 'initial-style:0;\n' : 'initial-style:1;\n') + gabcdata;
	} else if (tags.has('invitatorium')) {
    gabcdata = 'initial-style:0;\n' + gabcdata;
  }

  gabc = gabc.replace(/<v>\\([VRA])bar<\/v>/g,function(match,barType) {
    return barType + '/.';
  }).replace(/(<b>[^<]+)<sp>'(?:oe|œ)<\/sp>/g,'$1œ</b>\u0301<b>') // character doesn't work in the bold version of this font.
  .replace(/<b><\/b>/g,'')
  .replace(/<sp>'(?:ae|æ)<\/sp>/g,'ǽ')
  .replace(/<sp>'(?:oe|œ)<\/sp>/g,'œ́')
  .replace(/<v>\\greheightstar<\/v>/g,'*')
  .replace(/<\/?i>/g,'_')
  .replace(/<\/?nlba>/g,'');
  return gabcdata + gabc;
}

